A42-xds-ring-hash-lb-policy.md

```diff
@@ -0,0 +1,580 @@
+
+Here is how gRPC will handle each type of hash policy:
+
+- [`header`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L710):
+  gRPC will support this type of hash policy.  This allows hashing on a
+  request header.
+- [`cookie`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L713):
+  gRPC does not support HTTP cookies, so we will not support this hash
+  policy.  If specified, this policy will not return any result.
+- [`connection_properties`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L716):
+  In Envoy, this allows hashing based on the source IP address, which
+  ensures that all requests from a given client are sent to the same
+  endpoint.  However, gRPC cannot support this, because it is a client
+  and not a proxy, so it does not know what source address will be used
+  for the request until after it has finished load balancing.  If
+  specified, this policy will not return any result.  (However, see
+  `filter_state` below for an alternative.)
+- [`query_parameter`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L719):
+  gRPC does not support query parameters, so we will not support this
+  hash policy.  If specified, this policy will not return any result.
+- [`filter_state`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L722):
+  gRPC does not currently support the general-purpose concept of filter state
+  the way that Envoy does.  However, we will support one special filter state
+  [`key`](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/api/envoy/config/route/v3/route_components.proto#L703)
+  called `io.grpc.channel_id`, which will hash to the same value for all
+  requests on a given gRPC channel.  This can be used in similar situations
+  to where Envoy uses `connection_properties` to hash on the source IP address.
+
+#### XdsClient Changes
+
+In gRPC, the `xds` resolver registers a watcher with the `XdsClient` object
+to get the `RouteConfiguration` resource.  The struct returned by that
+watcher will be extended to include the following fields in each route
+(C++ syntax):
+
+```
+struct HashPolicy {
+  enum Type { HEADER, CHANNEL_ID };
+  Type type;
+  bool terminal = false;
+  // Fields used for type HEADER.
+  std::string header_name;
+  std::unique_ptr<RE2> regex;
+  std::string regex_substitution;
+};
+std::vector<HashPolicy> hash_policies;
+```
+
+The `XdsClient` will populate these fields from the fields of the
+`RouteAction` message described above.
+
+#### `XdsConfigSelector` Changes
+
+The `XdsConfigSelector` is created by the xds resolver and is responsible
```

Comment:
Should we mention somewhere here that xxhash is used for this step?

Reply:
Done.